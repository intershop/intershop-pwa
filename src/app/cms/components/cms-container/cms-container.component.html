<div [ngClass]="containerClasses" class="content-container">
  <ish-slot-wrapper [pagelet]="pagelet" slot="app_sf_responsive_cm:slot.container.content.pagelet2-Slot">
    <ish-content-pagelet *ngFor="let pagelet of contentSlotPagelets" [pagelet]="pagelet"></ish-content-pagelet>
  </ish-slot-wrapper>
</div>

<!-- TODO: (CMS Component Configuration)
  TrackingCode
  TrackingScript
-->
