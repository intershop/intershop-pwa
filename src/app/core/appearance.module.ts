import { NgModule } from '@angular/core';
import { NgbPopoverConfig } from '@ng-bootstrap/ng-bootstrap';
import { ToastrModule } from 'ngx-toastr';

import { IconModule } from './icon.module';
import { ThemeService } from './utils/theme/theme.service';

@NgModule({
  imports: [
    IconModule,
    ToastrModule.forRoot({
      closeButton: true,
      timeOut: 3000,
      positionClass: 'toast-top-full-width', // toast-top-center
      preventDuplicates: true,
    }),
  ],
})
export class AppearanceModule {
  constructor(popoverConfig: NgbPopoverConfig, themeService: ThemeService) {
    popoverConfig.placement = 'top';
    popoverConfig.triggers = 'hover';
    popoverConfig.container = 'body';

    themeService.init();
  }
}
