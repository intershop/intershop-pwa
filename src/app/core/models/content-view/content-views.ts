// tslint:disable:project-structure
// due to recursive depending types the definitions must remain in the same file
import { memoize, once } from 'lodash-es';

import { ContentPageletEntryPoint } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.model';
import { ContentConfigurationParameters } from '../content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentPagelet } from '../content-pagelet/content-pagelet.model';
import { ContentSlot } from '../content-slot/content-slot.model';

export interface ContentConfigurationParameterView {
  hasParam(key: string): boolean;
  stringParam(key: string, defaultValue?: string): string;
  numberParam(key: string, defaultValue?: number): number;
  booleanParam(key: string, defaultValue?: boolean): boolean;
  configParam<T extends object>(key: string): T;
}

export interface ContentEntryPointView {
  id: string;
  domain: string;
  displayName?: string;
  pagelets(): ContentPageletView[];
}

export interface ContentSlotView extends ContentEntryPointView, ContentConfigurationParameterView {}

export interface ContentPageletEntryPointView extends ContentEntryPointView, ContentConfigurationParameterView {
  id: string;
  domain: string;
  displayName: string;
  resourceSetId: string;
  pagelets(): ContentPageletView[];
}

export interface ContentPageletView extends ContentConfigurationParameterView {
  definitionQualifiedName: string;
  configurationParameters: ContentConfigurationParameters;
  id: string;
  domain: string;
  displayName: string;
  slot(qualifiedName: string): ContentSlotView;
}

const paramMemoize = (key, defaultValue) => JSON.stringify({ key, defaultValue });

export const createContentConfigurationParameterView = (
  params: ContentConfigurationParameters
): ContentConfigurationParameterView => ({
  hasParam: memoize(key => Object.keys(params).includes(key)),
  booleanParam: memoize(
    (key, defaultValue = undefined) =>
      Object.keys(params).includes(key)
        ? typeof params[key] === 'string' && (params[key] as string).toLowerCase().trim() === 'true'
        : defaultValue,
    paramMemoize
  ),
  stringParam: memoize(
    (key, defaultValue = undefined) =>
      Object.keys(params).includes(key)
        ? typeof params[key] === 'string'
          ? (params[key] as string)
          : JSON.stringify(params[key])
        : defaultValue,
    paramMemoize
  ),
  numberParam: memoize((key, defaultValue = NaN) => Number(params[key]) || defaultValue, paramMemoize),
  configParam: <T extends object>(key) => params[key] as T,
});

export const createContentPageletView = (
  id: string,
  pagelets: { [id: string]: ContentPagelet }
): ContentPageletView => {
  const pagelet = pagelets[id];
  return {
    definitionQualifiedName: pagelet.definitionQualifiedName,
    configurationParameters: pagelet.configurationParameters,
    id: pagelet.id,
    domain: pagelet.domain,
    displayName: pagelet.displayName,
    slot: !pagelet.slots
      ? () => undefined
      : memoize(qualifiedName =>
          createContentSlotView(
            pagelet,
            pagelet.slots.find(slot => slot.definitionQualifiedName === qualifiedName),
            pagelets
          )
        ),
    ...createContentConfigurationParameterView(pagelet.configurationParameters || {}),
  };
};

export const createContentEntryPointView = (
  id: string,
  domain: string,
  pageletIDs: string[],
  pagelets: { [id: string]: ContentPagelet }
): ContentEntryPointView => ({
  id,
  domain,
  pagelets:
    pageletIDs && pageletIDs.length
      ? once(() => pageletIDs.map(pId => createContentPageletView(pId, pagelets)))
      : () => [],
});

export const createContentSlotView = (
  pagelet: ContentPagelet,
  slot: ContentSlot,
  pagelets: { [id: string]: ContentPagelet }
): ContentSlotView =>
  !slot
    ? undefined
    : {
        ...createContentEntryPointView(slot.definitionQualifiedName, pagelet.domain, slot.pageletIDs, pagelets),
        ...createContentConfigurationParameterView(slot.configurationParameters || {}),
        displayName: slot.displayName,
      };

export const createContentPageletEntryPointView = (
  pageletEntryPoint: ContentPageletEntryPoint,
  pagelets: { [id: string]: ContentPagelet }
): ContentPageletEntryPointView => ({
  id: pageletEntryPoint.id,
  domain: pageletEntryPoint.domain,
  resourceSetId: pageletEntryPoint.resourceSetId,
  displayName: pageletEntryPoint.displayName,
  ...createContentEntryPointView(
    pageletEntryPoint.id,
    pageletEntryPoint.domain,
    pageletEntryPoint.pageletIDs,
    pagelets
  ),
  ...createContentConfigurationParameterView(pageletEntryPoint.configurationParameters || {}),
});
