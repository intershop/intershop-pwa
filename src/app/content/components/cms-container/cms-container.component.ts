import { ChangeDetectionStrategy, Component, Input, OnInit } from '@angular/core';

import { ContentPageletHelper } from '../../../models/content-pagelet/content-pagelet.helper';
import { ContentPagelet } from '../../../models/content-pagelet/content-pagelet.model';

// tslint:disable-next-line:project-structure
@Component({
  selector: 'ish-cms-container',
  templateUrl: './cms-container.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CMSContainerComponent implements OnInit {
  @Input()
  pagelet: ContentPagelet;

  contentSlotPagelets: ContentPagelet[] = [];
  containerClasses = '';

  getConfigurationParameterValue = ContentPageletHelper.getConfigurationParameterValue;

  ngOnInit() {
    const upperBound = this.getConfigurationParameterValue(this.pagelet, 'UpperBound', 'number');
    this.contentSlotPagelets = upperBound
      ? this.pagelet.slots['app_sf_responsive_cm:slot.container.content.pagelet2-Slot'].pagelets.slice(0, upperBound)
      : this.pagelet.slots['app_sf_responsive_cm:slot.container.content.pagelet2-Slot'].pagelets;

    this.containerClasses = this.getGridCSS(this.getConfigurationParameterValue(this.pagelet, 'Grid'));
    this.containerClasses += this.getConfigurationParameterValue(this.pagelet, 'CSSClass')
      ? this.getConfigurationParameterValue(this.pagelet, 'CSSClass')
      : '';
  }

  getGridCSS(grid: string): string {
    let gridCSS = '';

    // transform an incomming string like "ExtraSmall:12,Small:6,Medium:4,Large:0" to a grid object
    const gridObject = { ExtraSmall: 0, Small: 0, Medium: 0, Large: 0 };
    grid.split(',').map(element => {
      gridObject[element.split(':')[0]] = Number(element.split(':')[1]);
    });

    // the 'hidden-__' classes replacement is not 100% compatible to Bootstrap 3
    if (gridObject.ExtraSmall !== 0) {
      gridCSS += gridObject.ExtraSmall !== -1 ? `col-${gridObject.ExtraSmall} ` : 'd-none d-md-block ';
    }
    if (gridObject.Small !== 0) {
      gridCSS += gridObject.Small !== -1 ? `col-md-${gridObject.Small} ` : 'd-md-none d-lg-block ';
    }
    if (gridObject.Medium !== 0) {
      gridCSS += gridObject.Medium !== -1 ? `col-lg-${gridObject.Medium} ` : 'd-lg-none d-xl-block ';
    }
    if (gridObject.Large !== 0) {
      gridCSS += gridObject.Large !== -1 ? `col-xl-${gridObject.Large} ` : 'd-xl-none ';
    }

    // 'float-left' is added to get the similar styling as with Bootstrap 3
    if (gridCSS !== '') {
      gridCSS += 'float-left ';
    }

    return gridCSS;
  }
}
