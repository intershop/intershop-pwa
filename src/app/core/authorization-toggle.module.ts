import { ModuleWithProviders, NgModule } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';

import { AuthorizationToggleDirective } from './directives/authorization-toggle.directive';
import { AuthorizationToggleService, checkPermission } from './utils/authorization-toggle/authorization-toggle.service';
import { whenTruthy } from './utils/operators';

@NgModule({
  declarations: [AuthorizationToggleDirective],
  exports: [AuthorizationToggleDirective],
})
export class AuthorizationToggleModule {
  private static permissions = new ReplaySubject<string[]>(1);

  static forTesting(...permissions: string[]): ModuleWithProviders<AuthorizationToggleModule> {
    AuthorizationToggleModule.switchTestingPermissions(...permissions);
    return {
      ngModule: AuthorizationToggleModule,
      providers: [
        {
          provide: AuthorizationToggleService,
          useValue: {
            isAuthorizedTo: (permission: string) =>
              AuthorizationToggleModule.permissions.pipe(
                whenTruthy(),
                map(perms => checkPermission(perms, permission))
              ),
          },
        },
      ],
    };
  }

  static switchTestingPermissions(...permissions: string[]) {
    AuthorizationToggleModule.permissions.next(permissions);
  }
}

export { AuthorizationToggleService } from './utils/authorization-toggle/authorization-toggle.service';
export { AuthorizationToggleGuard } from './guards/authorization-toggle.guard';
