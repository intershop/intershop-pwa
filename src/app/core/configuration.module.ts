import { NgModule } from '@angular/core';

import { SPECIAL_HTTP_ERROR_HANDLER } from './interceptors/icm-error-mapper.interceptor';
import { createPaymentErrorHandler } from './utils/http-error/create-payment.error-handler';
import { editPasswordErrorHandler } from './utils/http-error/edit-password.error-handler';
import { LoginUserErrorHandler } from './utils/http-error/login-user.error-handler';
import { requestReminderErrorHandler } from './utils/http-error/request-reminder.error-handler';
import { updatePasswordErrorHandler } from './utils/http-error/update-password.error-handler';

@NgModule({
  providers: [
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: updatePasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useClass: LoginUserErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: requestReminderErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: editPasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: createPaymentErrorHandler, multi: true },
  ],
})
export class ConfigurationModule {}
