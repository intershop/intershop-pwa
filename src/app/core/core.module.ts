import { APP_BASE_HREF, PlatformLocation } from '@angular/common';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { ErrorHandler, NgModule, Optional, SkipSelf } from '@angular/core';
import { ServiceWorkerModule } from '@angular/service-worker';
import { TranslateModule } from '@ngx-translate/core';

import { AppearanceModule } from './appearance.module';
import { ConfigurationModule } from './configuration.module';
import { FeatureToggleModule } from './feature-toggle.module';
import { IdentityProviderModule } from './identity-provider.module';
import { ICMErrorMapperInterceptor } from './interceptors/icm-error-mapper.interceptor';
import { IdentityProviderInterceptor } from './interceptors/identity-provider.interceptor';
import { MockInterceptor } from './interceptors/mock.interceptor';
import { InternationalizationModule } from './internationalization.module';
import { StateManagementModule } from './state-management.module';
import { DefaultErrorHandler } from './utils/default-error-handler';

@NgModule({
  imports: [
    AppearanceModule,
    ConfigurationModule,
    FeatureToggleModule,
    HttpClientModule,
    IdentityProviderModule,
    InternationalizationModule,
    ServiceWorkerModule.register('ngsw-worker.js', { enabled: SERVICE_WORKER }),
    StateManagementModule,
  ],
  providers: [
    { provide: HTTP_INTERCEPTORS, useClass: ICMErrorMapperInterceptor, multi: true },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: IdentityProviderInterceptor,
      multi: true,
    },
    { provide: HTTP_INTERCEPTORS, useClass: MockInterceptor, multi: true },
    { provide: ErrorHandler, useClass: DefaultErrorHandler },
    {
      provide: APP_BASE_HREF,
      useFactory: (s: PlatformLocation, baseHref: string) => baseHref || s.getBaseHrefFromDOM(),
      deps: [PlatformLocation, [new Optional(), new SkipSelf(), APP_BASE_HREF]],
    },
  ],
  // exports needed to use the cookie banner in the AppComponent
  exports: [TranslateModule],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
    if (parentModule) {
      throw new Error('CoreModule is already loaded. Import it in the AppModule only');
    }
  }
}
