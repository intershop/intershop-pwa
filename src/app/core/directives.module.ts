import { NgModule } from '@angular/core';

import { ClickOutsideDirective } from './directives/click-outside.directive';
import { IdentityProviderCapabilityDirective } from './directives/identity-provider-capability.directive';
import { IntersectionObserverDirective } from './directives/intersection-observer.directive';
import { ProductContextAccessDirective } from './directives/product-context-access.directive';
import { ProductContextDirective } from './directives/product-context.directive';
import { ServerHtmlDirective } from './directives/server-html.directive';

@NgModule({
  declarations: [
    ClickOutsideDirective,
    IdentityProviderCapabilityDirective,
    IntersectionObserverDirective,
    ProductContextAccessDirective,
    ProductContextDirective,
    ServerHtmlDirective,
  ],
  exports: [
    ClickOutsideDirective,
    IdentityProviderCapabilityDirective,
    IntersectionObserverDirective,
    ProductContextAccessDirective,
    ProductContextDirective,
    ServerHtmlDirective,
  ],
})
export class DirectivesModule {}
