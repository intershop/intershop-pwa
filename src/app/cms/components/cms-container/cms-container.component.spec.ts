import { ComponentFixture, TestBed, async } from '@angular/core/testing';

import { createContentPageletView } from 'ish-core/models/content-view/content-views';
import { MockComponent } from 'ish-core/utils/dev/mock.component';

import { CMSContainerComponent } from './cms-container.component';

describe('Cms Container Component', () => {
  let component: CMSContainerComponent;
  let fixture: ComponentFixture<CMSContainerComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        CMSContainerComponent,
        MockComponent({
          selector: 'ish-content-pagelet',
          template: 'outlet',
          inputs: ['pagelet'],
        }),
        MockComponent({
          selector: 'ish-content-slot',
          template: 'Slot',
          inputs: ['pagelet', 'definitionQualifiedName'],
        }),
        MockComponent({
          selector: 'ish-slot-wrapper',
          template: 'outlet',
          inputs: ['pagelet', 'definitionQualifiedName'],
        }),
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CMSContainerComponent);
    component = fixture.componentInstance;
    const slide1 = {
      definitionQualifiedName: 'fq',
      id: 'slide1',
      domain: 'domain',
      displayName: 'slide1',
    };
    const slide2 = {
      definitionQualifiedName: 'fq',
      id: 'slide2',
      domain: 'domain',
      displayName: 'slide2',
    };
    const pagelet = {
      definitionQualifiedName: 'fq',
      id: 'id',
      domain: 'domain',
      displayName: 'pagelet',
      configurationParameters: {
        CSSClass: 'foo-class',
        Grid: 'ExtraSmall:12,Small:6,Medium:4,Large:0',
      },
      slots: [
        {
          definitionQualifiedName: 'app_sf_responsive_cm:slot.container.content.pagelet2-Slot',
          pageletIDs: [slide1.id, slide2.id],
        },
      ],
    };
    component.pagelet = createContentPageletView(pagelet.id, {
      [pagelet.id]: pagelet,
      [slide1.id]: slide1,
      [slide2.id]: slide2,
    });
    element = fixture.nativeElement;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => component.ngOnChanges()).not.toThrow();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });
});
