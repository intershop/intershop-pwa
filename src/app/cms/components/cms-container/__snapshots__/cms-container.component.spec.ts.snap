// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cms Container Component should be created 1`] = `
<div
  id="root0"
  ng-version="7.1.4"
>
  <div
    class="content-container col-12 col-md-6 col-lg-4 float-left foo-class"
    ng-reflect-klass="content-container"
    ng-reflect-ng-class="col-12 col-md-6 col-lg-4 float"
  >
    <ish-slot-wrapper
      ng-reflect-pagelet="[object Object]"
      slot="app_sf_responsive_cm:slot.container.content.pagelet2-Slot"
    >
      outlet
    </ish-slot-wrapper>
  </div>
</div>
`;
